File: data structures and algorithms/java/08_019.java
import java.util.*;

public class RateLimiter {
    private final int messageLimit;
    private final int timeWindow;
    private final Map<String, Deque<Integer>> userMessages;
    
    public RateLimiter(int messageLimit, int timeWindow) {
        this.messageLimit = messageLimit;
        this.timeWindow = timeWindow;
        this.userMessages = new HashMap<>();
    }
    
    public boolean canSend(String userId, int timestamp) {
        // Initialize queue for new users
        if (!userMessages.containsKey(userId)) {
            userMessages.put(userId, new LinkedList<>());
        }
        
        Deque<Integer> messageQueue = userMessages.get(userId);
        
        // Remove messages outside the current time window
        while (!messageQueue.isEmpty() && 
               messageQueue.peekFirst() <= timestamp - timeWindow) {
            messageQueue.pollFirst();
        }
        
        // Check if adding a new message would exceed the limit
        if (messageQueue.size() < messageLimit) {
            messageQueue.addLast(timestamp);
            return true;
        }
        
        return false;
    }
}

File: data structures and algorithms/cs/08_019.cs
using System;
using System.Collections.Generic;

public class RateLimiter
{
   private readonly int messageLimit;
   private readonly int timeWindow;
   private readonly Dictionary<string, Queue<int>>
      userMessages;

   public RateLimiter(int messageLimit, int timeWindow)
   {
      this.messageLimit = messageLimit;
      this.timeWindow = timeWindow;
      this.userMessages = new Dictionary<string,
         Queue<int>>();
   }

   public bool CanSend(string userId, int timestamp)
   {
      if (!userMessages.ContainsKey(userId))
      {
         userMessages[userId] = new Queue<int>();
      }

      var messageQueue = userMessages[userId];

      while (messageQueue.Count > 0 &&
         messageQueue.Peek() <= timestamp - timeWindow)
      {
         messageQueue.Dequeue();
      }

      if (messageQueue.Count < messageLimit)
      {
         messageQueue.Enqueue(timestamp);
         return true;
      }

      return false;
   }
}